lib/client.js

```diff
@@ -114,6 +115,19 @@ module.exports = class Client extends EventEmitter {
       const contentEncoding = headers['content-encoding']
       const statusCode = res.statusCode
 
+      // "If HTTP 307, follow the redirect"
+      if (statusCode === 307) {
+        const redirectURL = new URL(headers['location'])
+        return this._makeRequest(Object.assign(
+          Object.assign({}, opts),
+          {
+            path: redirectURL.pathname,
+            hostname: redirectURL.hostname,
+            port: redirectURL.port,
+            protocol: redirectURL.protocol
+          }), backoff, callback_)
```

Comment:
`Object.assign` can take more than two arguments: `Object.assign({}, opts, { path: .. }`).

More importantly though, I think this won't work with how the query stream works atm, which has its own request options object. I'll see if I can refactor that a bit.

Reply:
Actually we're good, because in the case of HTTP 307 "future requests should still use the original URI." ([wikipedia](https://en.wikipedia.org/wiki/HTTP_307)). In addition, wherever we're redirected to, is also the party that will send us a `nextUri` if any (which may result in an effectively permanent redirect).

Reply:
So with the `Object.assign()` tweak, and not following redirects from HTTPS to HTTP, and being consistent with other internals (where `port` is a number) and compatible with #20, I suggest:

```js
return this._makeRequest(
  Object.assign({}, opts, {
    // Follow host redirects, but don't switch protocols.
    hostname: redirectURL.hostname,
    path: redirectURL.pathname + redirectURL.search,
    port: redirectURL.port ? parseInt(redirectURL.port, 10) : undefined
  }),
  backoff,
  callback_
)
```

Reply:
@vweevers thanks for the quick reply! (I'm glad to find this one is being mantained)
I'm updating it to do 1 Object.assign, re. the redirect behavior, are you sure protocol can't switch? 
because from wikipedia I interpreted only the request method can't change.

Reply:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/307
> The only difference between 307 and 302 is that 307 guarantees that the method and the body will not be changed when the redirected request is made.

Reply:
The protocol can switch, I just prefer it not to. If you're using a HTTPS endpoint, you wouldn't want it to drop down to HTTP.

Reply:
got it, I could check and only switch if the destination is HTTPS?

Reply:
> I could check and only switch if the destination is HTTPS?

I think that's too complicated for little gain - unless you know of a use case where HTTP/HTTPS endpoints are actually mixed (which doesn't sound ideal).

Reply:
makes sense. then if you don't mind I'll throw an error if the protocol is different, as we shouldn't hit the target location without using the protocol in Locaiton (that could be a security vulnerability)
eg:
``` js
        if (opts.protocol !== redirectURL.protocol) {
          throw new Error('307 Redirect protocol switch is not allowed');
        }
```
wdyt?

Reply:
Oh, you mean when the initial protocol is HTTP but we're redirected to HTTPS? Good point.

An error sounds good. Rather than throwing though, can we call the callback? Because of the asynchronicity.

Reply:
got it 👍

Reply:
added